linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,134 @@
+
+## Exposed API
+
+```C
+typedef struct COMPLETION_PORT_TAG* COMPLETION_PORT_HANDLE;
+
+#define COMPLETION_PORT_EPOLL_ACTION_VALUES \
+    COMPLETION_PORT_EPOLL_EPOLLRDHUP, \
+    COMPLETION_PORT_EPOLL_EPOLLIN, \
+    COMPLETION_PORT_EPOLL_EPOLLOUT, \
+    COMPLETION_PORT_EPOLL_ABANDONED, \
+    COMPLETION_PORT_EPOLL_ERROR
+
+MU_DEFINE_ENUM(COMPLETION_PORT_EPOLL_ACTION, COMPLETION_PORT_EPOLL_ACTION_VALUES)
+
+typedef void (*ON_COMPLETION_PORT_EVENT_COMPLETE)(void* context, COMPLETION_PORT_EPOLL_ACTION epoll_action);
+
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, int, completion_port_add, COMPLETION_PORT_HANDLE, completion_port, int, epoll_op, SOCKET_HANDLE, socket,
+    ON_COMPLETION_PORT_EVENT_COMPLETE, event_callback, void*, event_callback_ctx);
+MOCKABLE_FUNCTION(, void, completion_port_remove, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+### completion_port_create
+
+```C
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+```
+
+`completion_port_create` creates and initializes the completion port module
+
+`completion_port_create` shall allocate memory for a completion port object.
+
+`completion_port_create` shall create the epoll instance by calling `epoll_create`.
+
+`completion_port_create` shall create a thread that runs `epoll_worker_func` to handle the epoll events.
+
+On success `completion_port_create` shall return the allocated `COMPLETION_PORT_HANDLE`.
+
+If there are any errors then `completion_port_create` shall fail and return `NULL`.
+
+### completion_port_inc_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_inc_ref` handles the incrementing of the reference count of the module
+
+If `completion_port` is `NULL`, `completion_port_inc_ref` shall return.
+
+Otherwise `completion_port_inc_ref` shall increment the internally maintained reference count.
+
+### completion_port_dec_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_dec_ref` handles the decrement of the reference count and freeing the memory if necessary
+
+`completion_port_dec_ref` shall decrement the reference count for `completion_port`.
+
+If the reference count reaches 0, `completion_port_dec_ref` shall do the following:
+
+- increment the flag signaling that the threads can complete.
+
+- close the epoll object.
+
+- close the thread by calling `ThreadAPI_Join`
+
+- then the memory associated with `completion_port` shall be freed.
+
+### completion_port_add
+
+```C
+MOCKABLE_FUNCTION(, int, completion_port_add, COMPLETION_PORT_HANDLE, completion_port, int, epoll_op, SOCKET_HANDLE, socket,
```

Comment:
>completion_port_add [](http://example.com/codeflow?start=25&length=19)

Can completion_port_add be called while dec_ref is doing the destruction of the completion port?

Reply:
I think I'll need to add a pending call count for this.